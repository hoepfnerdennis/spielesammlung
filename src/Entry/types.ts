export type EntryProps = {
  name: string;
  description: string;
  players: {
    from: string;
    to: string;
  };
  age: string;
  duration: string;
};
