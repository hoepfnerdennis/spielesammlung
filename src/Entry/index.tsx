import React, { memo, createRef, useEffect } from 'react';
import styles from './styles.module.css';
import { IGame } from '../Store/types';
import useIntersection from './intersection';
import Feature from '../Features';

const Entry: React.SFC<IGame> = ({
  name,
  description,
  playersFrom,
  playersTo,
  age,
  duration,
  favorite,
  simpleRules,
  image,
  drinkingGame,
}): JSX.Element => {
  const entryRef = createRef<HTMLDivElement>();
  const intersecting = useIntersection(entryRef, true);
  const intersectingWithMargin = useIntersection(entryRef, true, { rootMargin: '200px' });

  useEffect(() => {
    if (intersectingWithMargin) {
      const loadImage = new Image();
      loadImage.src = image;
    }
  }, [intersectingWithMargin, image]);

  const player = playersFrom === playersTo ? playersFrom : `${playersFrom} - ${playersTo}`;

  return (
    <div className={styles.border} ref={entryRef}>
      {intersecting && (
        <div className={`${styles.container} ${favorite ? styles.fav : ''}`}>
          <div className={styles.imageContainer}>
            <img src={image} alt={`Bild von ${name}`} className={styles.image} />
          </div>
          <div className={styles.info}>
            <h2 className={styles.name}>{name}</h2>
            <p className={styles.description}>{description}</p>
            <Feature.List>
              <Feature.Item name="Anzahl Spieler" icon="👤" desc={`${player} Spieler`} />
              <Feature.Item name="Altersempfehlung" icon="📅" desc={`ab ${age} Jahren`} />
              <Feature.Item name="Spieldauer" icon="⏳" desc={duration} />
              <Feature.Item
                condition={simpleRules}
                name="Einfache Regeln"
                icon="🏋️‍♀️"
                desc="einfach"
              />
              <Feature.Item
                condition={drinkingGame}
                name="Trinkspiel"
                icon="🥂"
                desc="Trinkspiel"
              />
            </Feature.List>
          </div>
        </div>
      )}
    </div>
  );
};

export default memo(Entry);
