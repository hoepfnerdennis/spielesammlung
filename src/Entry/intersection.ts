import { useRef, useEffect, useState, useCallback } from 'react';

const useIntersection = (
  nodeRef: React.RefObject<HTMLElement>,
  triggerOnce = false,
  observerOptions?: IntersectionObserverInit
): [boolean] => {
  const observerRef = useRef<IntersectionObserver | undefined>();
  const [isIntersecting, setIsIntersecting] = useState<boolean>(false);

  // check for IntersectionObserver available
  if (typeof IntersectionObserver === 'undefined') {
    if (!isIntersecting) {
      setIsIntersecting(true);
    }
  }

  const observerCallback = useCallback(
    ([entry]: IntersectionObserverEntry[], observer: IntersectionObserver) => {
      // the first element is the element to observe
      if (entry && entry.isIntersecting) {
        setIsIntersecting(true);
        if (triggerOnce) {
          observer.unobserve(entry.target);
        }
      } else {
        setIsIntersecting(false);
      }
    },
    [triggerOnce]
  );

  // create new observer on mount and cleanup on unmount
  useEffect(() => {
    const currentNode = nodeRef.current;

    if (IntersectionObserver && currentNode) {
      observerRef.current = new IntersectionObserver(observerCallback, observerOptions);
      observerRef.current.observe(currentNode);
    }

    return (): void => {
      if (observerRef.current) {
        observerRef.current.disconnect();
        observerRef.current = undefined;
      }
    };
  }, [nodeRef, observerCallback, observerOptions]);

  return [isIntersecting];
};

export default useIntersection;
