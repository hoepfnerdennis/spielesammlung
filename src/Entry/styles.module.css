.border {
    max-width: 800px;
    margin: 0 auto;
}

.container {
    position: relative;
    
    margin: .6rem;
    padding: .6rem;

    border-radius: 3px;
    box-shadow: 0px 0px 10px 0px rgba(0, 0, 0, .3);

    display: flex;
    flex-direction: column;
    align-items: flex-start;
}

.imageContainer {
    flex-basis: 300px;
    height: 200px;
    margin-right: .6rem;
    align-self: center;
}

.image {
    object-fit: contain;
    width: 100%;
    height: 100%;

}

.info {
    flex: 1;
}

.name {
    margin: 0 0 .5rem 0;
    color: darkcyan;
}

.description {
    margin: 0 0 1rem 0;
}

.fav {
    border: 3px solid darkcyan;
}

@media (min-width: 600px) {
    .container {
        flex-direction: row;
    }
    .imageContainer {
        align-self: flex-start;
    }
}
