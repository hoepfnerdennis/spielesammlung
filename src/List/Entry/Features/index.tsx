import React from 'react';
import styles from './styles.module.css';

const Features: React.SFC<{
  playersFrom: number;
  playersTo: number;
  age: string;
  duration: string;
  simpleRules: boolean;
}> = ({ playersFrom, playersTo, age, duration, simpleRules }) => {
  return (
    <div className={styles.container}>
      <div className={styles.box}>
        <p className={styles.players}>
          <span role="img" aria-label="Anzahl Spieler">
            👤
          </span>
          <b className={styles.highlight}>
            {playersFrom === playersTo ? playersFrom : `${playersFrom} - ${playersTo}`} Spieler
          </b>
        </p>
        <p className={styles.players}>
          <span role="img" aria-label="Altersempfehlung">
            📅
          </span>
          <b className={styles.highlight}>ab {age} Jahren</b>
        </p>
        <p className={styles.players}>
          <span role="img" aria-label="Spieldauer">
            ⏳
          </span>
          <b className={styles.highlight}>{duration}</b>
        </p>
      </div>
      {simpleRules && (
        <div>
          <p className={styles.players}>
            <span role="img" aria-label="Einfache Regeln">
              🏋️‍♀️
            </span>
            <b className={styles.highlight}>einfach</b>
          </p>
        </div>
      )}
    </div>
  );
};

export default Features;
