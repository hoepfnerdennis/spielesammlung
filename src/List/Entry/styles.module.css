.border {
    max-width: 800px;
    margin: 0 auto;
}

.container {
    position: relative;
    
    margin: .6rem;
    padding: .6rem;

    border-radius: 3px;
    box-shadow: 0px 0px 10px 0px rgba(0, 0, 0, .3);

    display: flex;
    flex-direction: column;
    align-items: flex-start;
}

.image {
    flex-basis: 300px;

    margin-right: .6rem;
    object-fit: cover;

    height: 200px;
}

.info {
    flex: 1;
}

.name {
    margin: 0 0 .5rem 0;
    color: darkcyan;
}

.description {
    margin: 0 0 1rem 0;
}

.fav {
    border: 3px solid darkcyan;
}

@media (min-width: 600px) {
    .container {
        flex-direction: row;
    }
}
