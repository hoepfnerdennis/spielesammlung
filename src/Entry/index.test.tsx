import React from 'react';
import { render } from '@testing-library/react';
import Entry from './index';
import useIntersection from './intersection';
import getGame from '../../mockData/test-games';

jest.mock('../Features/index.tsx', () => ({
  List: ({ children }) => children,
  Item: ({ name, condition }) => <div data-condition={condition}>{name}</div>,
}));
jest.mock('./intersection.ts');

describe('Entry', () => {
  it.each([
    ['with intersection', true, getGame()],
    ['with favorite', true, getGame({ favorite: true })],
    ['with same players', true, getGame({ playersFrom: 2, playersTo: 2 })],
    ['default', false, getGame()],
  ])('should render %s', (_, intersecting, game) => {
    // @ts-ignore
    useIntersection.mockImplementation(() => intersecting);
    // eslint-disable-next-line react/jsx-props-no-spreading
    const { container } = render(<Entry {...game} />);
    expect(container).toMatchSnapshot();
  });
  it('should handle conditions correct', () => {
    // @ts-ignore
    useIntersection.mockImplementation(() => true);
    // eslint-disable-next-line react/jsx-props-no-spreading
    const { getByText } = render(<Entry {...getGame({ drinkingGame: true, simpleRules: true })} />);
    expect(getByText('Trinkspiel').dataset.condition).toBe('true');
    expect(getByText('Einfache Regeln').dataset.condition).toBe('true');
  });
});
