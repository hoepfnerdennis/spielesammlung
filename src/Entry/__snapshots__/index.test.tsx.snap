// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Entry should render default 1`] = `
<div>
  <div
    class="border"
  />
</div>
`;

exports[`Entry should render with favorite 1`] = `
<div>
  <div
    class="border"
  >
    <div
      class="container fav"
    >
      <div
        class="imageContainer"
      >
        <img
          alt="Bild von name"
          class="image"
          src="image"
        />
      </div>
      <div
        class="info"
      >
        <h2
          class="name"
        >
          name
        </h2>
        <p
          class="description"
        >
          description
        </p>
        <div>
          Anzahl Spieler
        </div>
        <div>
          Altersempfehlung
        </div>
        <div>
          Spieldauer
        </div>
        <div
          data-condition="false"
        >
          Einfache Regeln
        </div>
        <div
          data-condition="false"
        >
          Trinkspiel
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Entry should render with intersection 1`] = `
<div>
  <div
    class="border"
  >
    <div
      class="container "
    >
      <div
        class="imageContainer"
      >
        <img
          alt="Bild von name"
          class="image"
          src="image"
        />
      </div>
      <div
        class="info"
      >
        <h2
          class="name"
        >
          name
        </h2>
        <p
          class="description"
        >
          description
        </p>
        <div>
          Anzahl Spieler
        </div>
        <div>
          Altersempfehlung
        </div>
        <div>
          Spieldauer
        </div>
        <div
          data-condition="false"
        >
          Einfache Regeln
        </div>
        <div
          data-condition="false"
        >
          Trinkspiel
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Entry should render with same players 1`] = `
<div>
  <div
    class="border"
  >
    <div
      class="container "
    >
      <div
        class="imageContainer"
      >
        <img
          alt="Bild von name"
          class="image"
          src="image"
        />
      </div>
      <div
        class="info"
      >
        <h2
          class="name"
        >
          name
        </h2>
        <p
          class="description"
        >
          description
        </p>
        <div>
          Anzahl Spieler
        </div>
        <div>
          Altersempfehlung
        </div>
        <div>
          Spieldauer
        </div>
        <div
          data-condition="false"
        >
          Einfache Regeln
        </div>
        <div
          data-condition="false"
        >
          Trinkspiel
        </div>
      </div>
    </div>
  </div>
</div>
`;
